File: server-node-legacy/src/utils/hash.js
const crypto = require("crypto");

function normalizeAmountToCents(value) {
  const numeric = Number(value);
  if (!Number.isFinite(numeric) || numeric <= 0) {
    throw new Error("金额必须为大于 0 的数字");
  }

  return Math.round(numeric * 100);
}

function sha256Hex(payload) {
  return `0x${crypto.createHash("sha256").update(payload).digest("hex")}`;
}

function buildProjectPayload(project) {
  return [
    project.projectId,
    project.name.trim(),
    project.targetAmount,
    project.startTime,
    project.endTime,
    project.status
  ].join("|");
}

function buildDonationPayload(donation) {
  return [
    donation.donationId,
    donation.projectId,
    donation.isAnonymous ? "ANONYMOUS" : donation.donorName.trim(),
    donation.amount,
    donation.message ? donation.message.trim() : "",
    donation.donatedAt
  ].join("|");
}

function buildDisbursementPayload(disbursement) {
  return [
    disbursement.disbursementId,
    disbursement.projectId,
    disbursement.amount,
    disbursement.receiver.trim(),
    disbursement.purpose.trim(),
    disbursement.occurredAt
  ].join("|");
}

function verifyHash(payload, hash) {
  return sha256Hex(payload) === hash;
}

module.exports = {
  normalizeAmountToCents,
  sha256Hex,
  buildProjectPayload,
  buildDonationPayload,
  buildDisbursementPayload,
  verifyHash
};

File: server/internal/app/hash.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math"
	"strings"
)

func normalizeAmountToCents(value float64) (int64, error) {
	if math.IsNaN(value) || math.IsInf(value, 0) || value <= 0 {
		return 0, fmt.Errorf("金额必须为大于 0 的数字")
	}
	return int64(math.Round(value * 100)), nil
}

func sha256Hex(payload string) string {
	sum := sha256.Sum256([]byte(payload))
	return "0x" + hex.EncodeToString(sum[:])
}

func buildProjectPayload(projectID int64, name string, targetAmount int64, startTime, endTime, status string) string {
	return strings.Join([]string{
		fmt.Sprintf("%d", projectID),
		strings.TrimSpace(name),
		fmt.Sprintf("%d", targetAmount),
		startTime,
		endTime,
		status,
	}, "|")
}

func buildDonationPayload(donationID, projectID int64, donorName string, isAnonymous bool, amount int64, message, donatedAt string) string {
	if isAnonymous {
		donorName = "ANONYMOUS"
	}
	return strings.Join([]string{
		fmt.Sprintf("%d", donationID),
		fmt.Sprintf("%d", projectID),
		strings.TrimSpace(donorName),
		fmt.Sprintf("%d", amount),
		strings.TrimSpace(message),
		donatedAt,
	}, "|")
}

func buildDisbursementPayload(disbursementID, projectID int64, amount int64, receiver, purpose, occurredAt string) string {
	return strings.Join([]string{
		fmt.Sprintf("%d", disbursementID),
		fmt.Sprintf("%d", projectID),
		fmt.Sprintf("%d", amount),
		strings.TrimSpace(receiver),
		strings.TrimSpace(purpose),
		occurredAt,
	}, "|")
}
